### PR TITLE
SpecAugment Layer added

### PR DESCRIPTION
**Pull request by Toku11**
I didn't add white noise layer because it has been removed

### CODE REVIEW

File: kapre/augmentation.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+        assert (freq_param is not None) or (time_param is not None), "at least one param value should be defined"
+        
+        idt = input_data_format
+        self.input_data_format = K.image_data_format() if idt == _CH_DEFAULT_STR else idt
+
+    def call(self, x, training=None):
+        
+        if training is None:
+            training = K.learning_phase()
```
Comment (keunwoochoi) [inline kapre/augmentation.py:54]: Do you know if this is necessary? I naively assumed if `training` argument is included in `call()` method, Keras would deal with everything correctly.

File: kapre/augmentation.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            
+        if self.input_data_format == 'channels_first':
+            assert x.shape[1] == 1, 'SpecAugment does not support 2D images yet'
```
Comment (keunwoochoi) [inline kapre/augmentation.py:57]: Can't we just extend it so that `SpecAug` is applied to all the channels? I'd like `Kapre` to be as versatile as possible.

Comment (keunwoochoi) [reply kapre/augmentation.py:57]: Btw in this kind of code, I'd prefer
```
ch_axis = 1 if self.input_data_format == 'channels_first' else 3
if K.int_shape(x)[ch_axis] != 1:
    raise RuntimeError('error msg, why it is broken, what was the expected value, what value it received, etc')
```

File: kapre/augmentation.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    # `SpecAugment`
+
+    ```python
+    kapre.augmentation.SpecAugment(time_param=10, freq_param=10, **kwargs)
+    ```
+
+    Add masking to input data and output it
+    Args:
+        x: Spectrogram
+        freq_param: Param of freq masking
```
Comment (keunwoochoi) [inline kapre/augmentation.py:24]: what are the types and shapes of these args? Some examples, too.

File: kapre/augmentation.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+from tensorflow.keras.layers import Layer
+from tensorflow.python.keras.utils import tf_utils
+
+__all__ = [
+    'SpecAugment'
+]
+
+class SpecAugment(Layer):
+    """
+    # `SpecAugment`
```
Comment (keunwoochoi) [inline kapre/augmentation.py:15]: The docstring format has changed, please have a look on the most up-to-date kapre code!

File: kapre/augmentation.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        **kwargs
+    ):
+        
+        super(SpecAugment, self).__init__(**kwargs)
+
+        backend.validate_data_format_str(input_data_format)
+
+        self.freq_param = freq_param
+        self.time_param = time_param
+        self.uses_learning_phase = True
```
Comment (keunwoochoi) [inline kapre/augmentation.py:43]: What does this line do?

File: kapre/augmentation.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        self.supports_masking = True
```
Comment (keunwoochoi) [inline kapre/augmentation.py:44]: What does this line do?

File: kapre/augmentation.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+        assert (freq_param is not None) or (time_param is not None), "at least one param value should be defined"
```
Comment (keunwoochoi) [inline kapre/augmentation.py:46]: Let's use `if: raise()` instead of `assert`! That's the new way I'm following recently (for a good reason)

File: kapre/backend.py [hunk 8]
```diff
@@ -210,3 +210,54 @@ def mu_law_decoding(signal_mu, quantization_channels):
         tf.math.sign(signal) * (tf.math.exp(tf.math.abs(signal) * tf.math.log1p(mu)) - 1.0) / mu
     )
     return signal
+
+
+def freq_mask(self, input_f, param, name='freq_mask'):
```
Comment (keunwoochoi) [inline kapre/backend.py:215]: (** no `self` in a function).

I think `freq_mask` and `time_mask` should be merged into one universal function. Maybe with a name like `random_masking_along_axis` or something, so..

```python
def random_masking_along_axis(x, axis, x_min=0, x_max=None, max_width=None): 
  if x_max is None:
    x_max = x.shape[axis]
  if max_width is None:
    max_width = (x_max - x_min) // 4  # 4 is an arbitrary choice though
  # do the work..
```

**Review by keunwoochoi — CHANGES_REQUESTED** [review]
Thanks for the initiative. Because the target branch already had `keras.augmentation`, there's a merge conflict but it wouldn't be anything too annoying.

File: kapre/augmentation.py (hunk 1 above)
Comment (Toku11) [reply kapre/augmentation.py:54]: It is training flag is used in smart_cond and could be modified on Keras Model Subclassing

File: kapre/backend.py (hunk 8 above)
Comment (Toku11) [reply kapre/backend.py:215]: Sorry, typo error when I was moving to backend, it was part of my class, Im trying to follow new Kapre but it has changed a lot

File: kapre/augmentation.py (hunk 1 above)
Comment (keunwoochoi) [reply kapre/augmentation.py:54]: Yup and I assume, from https://www.tensorflow.org/guide/keras/custom_layers_and_models#privileged_training_argument_in_the_call_method, we don't need this if statement.

Comment (Toku11) [reply kapre/augmentation.py:54]: I just followed TensorFlow practices  https://github.com/tensorflow/tensorflow/blob/b36436b087bd8e8701ef51718179037cccdfc26e/tensorflow/python/keras/layers/core.py#L206 , I can remove it though

Comment (keunwoochoi) [reply kapre/augmentation.py:54]: ha, this is interesting. never mind, let's keep it as it is to be safe.

Comment (keunwoochoi) [reply kapre/augmentation.py:54]: thanks for the context :) i'll maybe change my other code like that as well.

File: kapre/augmentation.py (hunk 2 above)
Comment (Toku11) [reply kapre/augmentation.py:57]: Well, I think I didn't describe the error correctly, It does support 2D images what It doesn't support is masking for images with depth >  1, but I'll work on that as soon as I have available time

**Comment by Toku11** [comment]
Ok, let me know if you need something else, I created a branch from master and there was no augmentation :D I didn't see 0.3.3 branch hh

File: kapre/augmentation.py (hunk 2 above)
Comment (keunwoochoi) [reply kapre/augmentation.py]: yeah but what's the reason behind the limitation? i think as a package, if kapre has a specaug, it should definitely support multi-channel inputs.

File: kapre/backend.py [hunk 9]
```diff
@@ -212,52 +212,26 @@ def mu_law_decoding(signal_mu, quantization_channels):
     return signal
 
 
-def freq_mask(self, input_f, param, name='freq_mask'):
-    """Apply masking to a spectrogram in the freq domain.
-    TensorFlow/io
-    
+def random_masking_along_axis(input_f, param, axis, name='masking'):
```
Comment (keunwoochoi) [inline kapre/backend.py:215]: as in the example i prototyped (
```
def random_masking_along_axis(x, axis, x_min=0, x_max=None, max_width=None): 
  if x_max is None:
    x_max = x.shape[axis]
  if max_width is None:
    max_width = (x_max - x_min) // 4  # 4 is an arbitrary choice though
  # do the work..
```)
I think this `param` should be unpacked and listed in this API.

File: kapre/backend.py [hunk 10]
```diff
@@ -212,52 +212,26 @@ def mu_law_decoding(signal_mu, quantization_channels):
...
+    """
+    Apply masking to a spectrogram in the time/freq domain.
     Args:
       input: An audio spectogram.
```
Comment (keunwoochoi) [inline kapre/backend.py:219]: name mistmatch. also, the format would be..
```
input (`Tensor`): Audio input spectrogram 
```

File: kapre/backend.py [hunk 11]
```diff
@@ -212,52 +212,26 @@ def mu_law_decoding(signal_mu, quantization_channels):
...
       param: Parameter of freq masking.
       name: A name for the operation (optional).
-
     Returns:
       A tensor of spectrogram.
     """
     # TODO: Support audio with channel > 1.
-    freq_max = tf.shape(input_f)[1]
+    _max = tf.shape(input_f)[axis + 1]
+    _shape = [-1, 1, 1, 1]
```
Comment (keunwoochoi) [inline kapre/backend.py:227]: I don't think we need to reshape the input tensor. We can instead keep the code cleaner/shorter by just passing the right `axis` argument in the layer implementation. At the same time, this function would just blindly do masking over the specified `axis`.

File: kapre/backend.py [hunk 12]
```diff
@@ -212,52 +212,26 @@ def mu_law_decoding(signal_mu, quantization_channels):
...
+    _shape[axis + 1] = _max
     f = tf.random.uniform(shape=(), minval=0, maxval=param, dtype=tf.dtypes.int32)
     f0 = tf.random.uniform(
-        shape=(), minval=0, maxval=freq_max - f, dtype=tf.dtypes.int32
+        shape=(), minval=0, maxval=_max - f, dtype=tf.dtypes.int32
     )
-    indices = tf.reshape(tf.range(freq_max), (-1,freq_max,1,1))
+    indices = tf.reshape(tf.range(_max), tuple(_shape))
```
Comment (keunwoochoi) [inline kapre/backend.py:233]: now because we're taking an arbitrary shaped input tensor, the reshaping should be taken care well. and that should be verified with `unittest`.

File: kapre/augmentation.py [hunk 13]
```diff
@@ -40,10 +45,9 @@ def __init__(
 
         self.freq_param = freq_param
```
Comment (keunwoochoi) [inline kapre/augmentation.py:53]: if the backend API changes, the API of this layer should follow. Maybe it should be like
```python
def __init__(self, time_min=0, time_max=None, time_max_width=None, freq_min=0, freq_max=None, freq_max_width=None, data_format='default', **kwargs):
```
where if `time_min` is `None`, it doesn't mask over time (and same for frequency axis).
Please note that all `Kapre` layers are compatible with both `channels_first` and `channels_last` data format, so this one should be, too.

**Review by keunwoochoi — CHANGES_REQUESTED** [review]
Thanks for the following up. I requested some changes. I acknowledge that this could be not a very little amount of work, but please understand that I have to be responsible for the consistency of the package. 
Besides the comments, some unit tests for the expected behaviors of the function and layer + save/load test of the layer is a mandatory, as done for other layers.. so that we can trust `Kapre`!

**Comment by bagustris** [comment]
Any progress on this PR? Having SpecAugment as Keras layer would be very useful for the DL-based audio processing community...

**Comment by Toku11** [comment]
> Any progress on this PR? Having SpecAugment as Keras layer would be very useful for the DL-based audio processing community...

Hi bagustris, unfortunately I was working in a project and I didn't have time to do proper test and documentation as per requested, however there is a functional version, I would really appreciate if you can continue with the work and make a pull request in the repository, please check my code https://github.com/Toku11/kapre/blob/d64e19d4917a5c7f8f109b2cfe5b7e06d118b8e2/kapre/augmentation.py#L21

**Comment by MichaelisTrofficus** [comment]
Hi, it's a pity I hadn't seen this pull request before. A couple of days ago I uploaded a package to Pypi where I do just this, that is, a custom layer of tensorflow.keras that implements the SpecAugment technique. 

This is the repo if you want to take a look: [https://github.com/MichaelisTrofficus/spec_augment]()

If you see that it can be useful I can adapt it to kapre and make a new pull request with proper testing and documentation or simply continue this one but with my own implementation.
